### PR TITLE
Avoid immutable dictionary allocations when creating immutable dataflow analysis result

### PR DESCRIPTION
**Pull request by mavasani**
1. https://github.com/dotnet/roslyn-analyzers/commit/de41bc3c86148efcf6b34d0a113bb6f51b9b515f: Commit from https://github.com/dotnet/roslyn-analyzers/pull/1946 (unrelated to this PR)
2. https://github.com/dotnet/roslyn-analyzers/commit/9d08f124848323ab90571f9937bb8c398fd39dbc: Core commit that avoids immutable dictionary allocations when creating block analysis result. Instead we hold onto the final analysis data object in the block analysis result, expose them as sealed dictionary analysis data (so the analyzers do not accidentally modify the computed result), and finally dispose them off during dispose invocation.
3. ~~https://github.com/dotnet/roslyn-analyzers/commit/fb505f51f61457ba75ad9cfbfbe5a846929d2510: Commit 2. above meant that now the analysis results exposed to analyzers are disposable. This commit updates the ported shipping FxCop dataflow rules to ensure these results are disposed. @dotpaul, once this PR is merged in, you may want to do the same for security analyzers performing flow analysis.~~

### CODE REVIEW

**Comment by mavasani** [comment]
retest prtest/Windows/NT/Debug please

**Comment by mavasani** [comment]
Making the dataflow analysis result disposable turned out to be too aggressive, preventing us from caching and reusing results across different analyzers/interprocedural invocations. I will revert the last commit and also revert the change to make `DataFlowAnalysisResult` disposable - I'll retain the finalizer for cleanup.

**Comment by mavasani** [comment]
@dotpaul This is now ready for review, and you don't need to make any changes for security rules.

**Comment by mavasani** [comment]
retest this please

**Comment by mavasani** [comment]
retest this please

File: src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractBlockAnalysisResult.cs [hunk 1]
```diff
@@ -1,17 +1,39 @@
...
+
+        public bool IsDisposed { get; private set; }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            IsDisposed = true;
+        }
+
+#pragma warning disable CA1063 // Implement IDisposable Correctly - We want to ensure that we cleanup managed resources even when object was not explicitly disposed.
+        ~AbstractBlockAnalysisResult()
```
Comment (sharwell) [inline src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractBlockAnalysisResult.cs]: :exclamation: let's have a discussion about this in our next design meeting

Comment (mavasani) [reply src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractBlockAnalysisResult.cs]: Sure, there is no rush to get this in and I also want to re-evaluate alternate ways to tackle the performance challenges of analyzing huge method bodies - for example, [CSharpCommandLineParser.Parse](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,293eb00017c0ffd4) generates a CFG with 500+ basic blocks and ~200 locals/parameters and generating/cloning dictionary(s) per basic block with 200+ key/value pairs and then merging data values across basic blocks is very expensive, especially when done iteratively till it reaches a fixed point. The current analysis performance is fine for reasonable sized methods.

**Comment by mavasani** [comment]
Closing this for now - short answer is any solution here should not use finalizers :-)

**Pull request closed without merging**